url.c

```diff
@@ -1463,6 +1463,122 @@ int url_mount(url_t *url, char *dir, int (*test_func)(url_t *))
 
 
 /*
+ * Warn if signature check failed and ask user what to do.
+ *
+ * Return 0 if it's ok to continue or 1 if we should report en error.
+ */
+int warn_signature_failed(char *file_name)
+{
+  int i, win, err = 0;
+  char *buf = NULL;
+
+  if(config.sig_failed) {
+    strprintf(&buf,
+      "%s: %s\n\n%s",
+      file_name,
+      config.sig_failed == 1 ? "File not signed." : "Invalid signature.",
+      config.sig_failed == 1 ? "If you really trust your repository, you may continue in an insecure mode." : "Installation aborted."
+    );
+    if(!(win = config.win)) util_disp_init();
+    if(config.sig_failed == 1) {
+      i = dia_okcancel(buf, NO);
+    }
+    else {
+      dia_message(buf, MSGTYPE_ERROR);
+      i = NO;
+    }
+    if(!win) util_disp_done();
+    if(i == YES) {
+      config.secure = 0;
+      config.sig_failed = 0;
+      err = 0;
+    }
+    else {
+      err = 1;
+    }
+  }
+
+  str_copy(&buf, NULL);
+
+  return err;
+}
+
+
+/*
+  Test if 'file' is a gpg signed file.
+  If so, unpack it (replacing 'file') and verify signature.
+  if 'check' is set, update config.sig_failed and show warning to user.
+
+  Return values:
+    -1: file or gpg not found
+     0: file gpg format, sig ok
+     1: file gpg format, sig wrong
+     2: file not gpg format
+*/
+int is_gpg_signed(char *file, int check)
+{
+  char *cmd = NULL, *buf = NULL;
+  int err = -1, is_sig = 0, sig_ok = 0;
+  size_t len = 0;
+  FILE *f;
+
+  if(util_check_exist(file) != 'r') {
+    if(config.debug) fprintf(stderr, "%s: gpg check = %d\n", file, err);
+
+    return err;
+  }
+
+  strprintf(&cmd,
+    "gpg --homedir /root/.gnupg --batch --no-default-keyring --keyring /installkey.gpg "
+    "--ignore-valid-from --ignore-time-conflict --output %s.unpacked %s 2>&1",
```

Comment:
file is a temporary file with a generated name; I've how put quotes around it anyway